File: crates/defi-protocols/src/liquid_staking/kinetiq.rs
use alloy::primitives::{Address, U256};
use alloy::providers::ProviderBuilder;
use alloy::sol;
use alloy::sol_types::SolCall;
use async_trait::async_trait;

use defi_core::error::{DefiError, Result};
use defi_core::traits::LiquidStaking;
use defi_core::types::*;

sol! {
    #[sol(rpc)]
    interface IKinetiq {
        function stake() external payable returns (uint256);
        function requestUnstake(uint256 amount) external returns (uint256);
        function exchangeRate() external view returns (uint256);
        function totalStaked() external view returns (uint256);
    }
}

pub struct Kinetiq {
    name: String,
    staking: Address,
    liquid_token: Address,
    rpc_url: Option<String>,
}

impl Kinetiq {
    pub fn new(name: String, staking: Address, liquid_token: Address) -> Self {
        Self {
            name,
            staking,
            liquid_token,
            rpc_url: None,
        }
    }

    pub fn from_contracts(
        name: String,
        contracts: &std::collections::HashMap<String, Address>,
        rpc_url: Option<String>,
    ) -> Result<Self> {
        let staking = contracts.get("staking").copied().ok_or_else(|| {
            DefiError::ContractError("Missing 'staking' contract address".to_string())
        })?;
        // kHYPE token address (same as staking contract for Kinetiq)
        let liquid_token = staking;
        Ok(Self {
            name,
            staking,
            liquid_token,
            rpc_url,
        })
    }

    fn rpc_url(&self) -> Result<url::Url> {
        self.rpc_url
            .as_ref()
            .ok_or_else(|| DefiError::RpcError("No RPC URL configured".to_string()))?
            .parse()
            .map_err(|e| DefiError::RpcError(format!("Invalid RPC URL: {e}")))
    }
}

#[async_trait]
impl LiquidStaking for Kinetiq {
    fn name(&self) -> &str {
        &self.name
    }

    async fn build_stake(&self, params: StakeParams) -> Result<DeFiTx> {
        let call = IKinetiq::stakeCall {};

        Ok(DeFiTx {
            description: format!("[{}] Stake {} HYPE for kHYPE", self.name, params.amount),
            to: self.staking,
            data: call.abi_encode().into(),
            value: params.amount, // Native HYPE sent as value
            gas_estimate: Some(200_000),
        })
    }

    async fn build_unstake(&self, params: UnstakeParams) -> Result<DeFiTx> {
        let call = IKinetiq::requestUnstakeCall {
            amount: params.amount,
        };

        Ok(DeFiTx {
            description: format!("[{}] Request unstake {} kHYPE", self.name, params.amount),
            to: self.staking,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(200_000),
        })
    }

    async fn get_info(&self) -> Result<StakingInfo> {
        let url = self.rpc_url()?;
        let provider = ProviderBuilder::new().connect_http(url);
        let contract = IKinetiq::new(self.staking, &provider);

        let exchange_rate = contract.exchangeRate().call().await.map_err(|e| {
            DefiError::RpcError(format!("[{}] exchangeRate failed: {e}", self.name))
        })?;

        let total_staked =
            contract.totalStaked().call().await.map_err(|e| {
                DefiError::RpcError(format!("[{}] totalStaked failed: {e}", self.name))
            })?;

        // Exchange rate is in 1e18 scale
        let rate_f64 = exchange_rate.to::<u128>() as f64 / 1e18;

        Ok(StakingInfo {
            protocol: self.name.clone(),
            staked_token: Address::ZERO, // Native HYPE
            liquid_token: self.liquid_token,
            exchange_rate: rate_f64,
            apy: None,
            total_staked,
        })
    }
}

File: ts/packages/defi-protocols/src/liquid_staking/kinetiq.ts
import { createPublicClient, http, parseAbi, encodeFunctionData, zeroAddress } from "viem";
import type { Address } from "viem";
import type { ILiquidStaking } from "@hypurrquant/defi-core";
import {
  DefiError,
  type ProtocolEntry,
  type StakeParams,
  type UnstakeParams,
  type StakingInfo,
  type DeFiTx,
} from "@hypurrquant/defi-core";

const KINETIQ_ABI = parseAbi([
  "function stake() external payable returns (uint256)",
  "function requestUnstake(uint256 amount) external returns (uint256)",
  "function totalStaked() external view returns (uint256)",
]);

const ORACLE_ABI = parseAbi([
  "function getAssetPrice(address asset) external view returns (uint256)",
]);

// WHYPE address on HyperEVM
const WHYPE: Address = "0x5555555555555555555555555555555555555555";
// HyperLend oracle address (Aave V3 compatible)
const HYPERLEND_ORACLE: Address = "0xc9fb4fbe842d57ea1df3e641a281827493a63030";

export class KinetiqAdapter implements ILiquidStaking {
  private readonly protocolName: string;
  private readonly staking: Address;
  private readonly liquidToken: Address;
  private readonly rpcUrl?: string;

  constructor(entry: ProtocolEntry, rpcUrl?: string) {
    this.protocolName = entry.name;
    this.rpcUrl = rpcUrl;
    const staking = entry.contracts?.["staking"];
    if (!staking) throw DefiError.contractError("Missing 'staking' contract address");
    this.staking = staking;
    this.liquidToken = entry.contracts?.["khype_token"] ?? staking;
  }

  name(): string {
    return this.protocolName;
  }

  async buildStake(params: StakeParams): Promise<DeFiTx> {
    const data = encodeFunctionData({ abi: KINETIQ_ABI, functionName: "stake" });
    return {
      description: `[${this.protocolName}] Stake ${params.amount} HYPE for kHYPE`,
      to: this.staking,
      data,
      value: params.amount,
      gas_estimate: 300_000,
    };
  }

  async buildUnstake(params: UnstakeParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: KINETIQ_ABI,
      functionName: "requestUnstake",
      args: [params.amount],
    });
    return {
      description: `[${this.protocolName}] Request unstake ${params.amount} kHYPE`,
      to: this.staking,
      data,
      value: 0n,
      gas_estimate: 300_000,
    };
  }

  async getInfo(): Promise<StakingInfo> {
    if (!this.rpcUrl) throw DefiError.rpcError("No RPC URL configured");
    const client = createPublicClient({ transport: http(this.rpcUrl) });

    const totalStaked = await client.readContract({
      address: this.staking,
      abi: KINETIQ_ABI,
      functionName: "totalStaked",
    }).catch((e: unknown) => {
      throw DefiError.rpcError(`[${this.protocolName}] totalStaked failed: ${e}`);
    }) as bigint;

    const [khypePrice, hypePrice] = await Promise.all([
      client.readContract({ address: HYPERLEND_ORACLE, abi: ORACLE_ABI, functionName: "getAssetPrice", args: [this.liquidToken] }).catch(() => 0n),
      client.readContract({ address: HYPERLEND_ORACLE, abi: ORACLE_ABI, functionName: "getAssetPrice", args: [WHYPE] }).catch(() => 0n),
    ]);

    const rateF64 =
      (hypePrice as bigint) > 0n && (khypePrice as bigint) > 0n
        ? Number((khypePrice as bigint) * 10n ** 18n / (hypePrice as bigint)) / 1e18
        : 1.0;

    return {
      protocol: this.protocolName,
      staked_token: zeroAddress as Address,
      liquid_token: this.liquidToken,
      exchange_rate: rateF64,
      total_staked: totalStaked,
    };
  }
}
